### DISCUSSION TITLE
Canon raw (.CR2) development options

### DISCUSSION THREAD

**Discussion by marek2023-1** [ImageMagick/ImageMagick#6500] [Digital Image Processing] [answered]
I'm trying to use magick (Win x64 Q16 portable) to convert .CR2 files into linear **unscaled** TIFFs (for deflickering timelapse). Previously I've used "dcraw -W -6 -T *.CR2" which works fine, but I'm struggling to replicate this in magick. I understand magick uses dcraw but I can't find how to set the dcraw options. I've managed to get linear TIFFs by using "-colorspace RGB" but the TIFFs are still scaled (white point at 65535). I've looked at the 'delegates'.xml' file, but that only refers to DNG raw files.
Any hints? Can magick do this? Googling around suggests that others do raw->linear unscaled TIFFs external to magick and than use magick to process the TIFFs. Is that the only way?

**Comment by snibgo** [comment]
IM is usually built with libraw. From the text output of "`magick -version`", look for `libraw` in the list of delegates.

See https://imagemagick.org/script/defines.php for defines you can set before reading the CR2 file. For example:
```
magick -define dng:no-auto-bright=true -define dng:output-color=0 abc.cr2 -set colorspace RGB -set gamma 1 out.tiff
```
This correctly makes a linear image. The options "`-set colorspace RGB -set gamma 1`" should both set the gamma to 1, so the metadata would record that the image is linear. However, this currently doesn't seem to work, suggesting a bug in IM:
```
magick out.tiff -verbose info: |grep Gamma
  Gamma: 0.454545
```

**Reply by marek2023-1** [reply]
Thanks for the quick reply, but it left me slightly more confused...
I check the delegate list (-version) and it doesn't include libraw (or for that matter dcraw). So I'm confused how magick converts my CR2 raw files.
I also checked the defines. They include` dng:no-auto-bright=true`, but I would have though that only affect conversion for DNG files, not CR2 files? Either way I'm getting my output TIFF files all intensity scaled (normalised? no-auto-bright=false?)

**Comment by snibgo** [comment]
What version of IM do you use? What is the text output from "`magick -version`"?

> I check the delegate list (-version) and it doesn't include libraw (or for that matter dcraw).

The `dng:` defines apply to all camera-raw files that are processed by `libraw`. ("DNG" is "Digital Negative". Camera-raw files include CR2, CR3, NEF, ARW and so on.) But you don't have `libraw`, so the defines don't help you.

That list shows internal delegates only. An *internal* delegate is built into the IM binaries. dcraw is an *external* delegate, meaning that IM calls it as an external program.

External delegates are called via `delegates.xml`. This will contain a line like:
```
  <delegate decode="dng:decode" stealth="True" command="dcraw.exe -6 -W -O &quot;%u.ppm&quot; &quot;%i&quot;"/>
```
You can edit this command to add or change or remove dcraw options. But don't change anything any of "`-O &quot;%u.ppm&quot; &quot;%i&quot;`"

**Reply by marek2023-1** [reply]
The output from -version is:
```
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193532217)
```
OK, I didn't understand the distinction between internal and external delegates. So far so educational, but I'd already played with delegates.xml and found that changing the 
`<delegate decode="dng:decode" stealth="True" command="dcraw.exe -6 -W -O &quot;%u.ppm&quot; &quot;%i&quot;"/>`
( e.g., changing the '-6' to '-4' or even deleting the line completely) made no difference to my TIFF output (still scaled intensity which it shouldn't be with '-4').

**Comment by snibgo** [comment, marked as answer]
Your internal delegates don't include `libraw` but do include `raw`, which is the same thing. So the defines should work for you. In particular, "`-define dng:no-auto-bright=true`" should work.

It does work for me. For example:
```
D:\temp>%IM7DEV%magick c:\web\im\AGA_1282.NEF -format %[fx:maxima] info:
1

D:\temp>%IM7DEV%magick -define dng:no-auto-bright=true c:\web\im\AGA_1282.NEF -format %[fx:maxima] info:
0.89320211
```
The output of `maxima` is on a nominal scale of 0.0 to 1.0. By default, it scales the image so the maximum is 1.0. The define prevents that.

If your CR2 file has clipped, there will be no difference.

It might help if you can link to your CR2 file, eg zip it and upload it here.

EDIT: You didn't show the first line of "`magick -version`". What is the version number, eg 7.1.1-5? If you have an old version, it might be before the `dng:` defines had an effect.

**Reply by marek2023-1** [reply]
Thank you, that seems to have solved it! For some reason I'd assumed that no-auto-bright defaults to true. It turns out that having it explicit in the command line **is** required.
For completeness I would add that I now can reproduce the effect of 'dcraw -w -o 1 -4 -T IMG_xxxx.CR2' with the script line:
`(IMG_xxxx.CR2 -colorspace RGB -define dng:no-auto-bright=true -define dng:use-camera-wb=true -gamma 1.0 -write IMG_xxxx.tiff)`
... At least as far as pixel values are concerned. There's still an issue with the TIFF embedded profile (if there is one). I thought that -define dng:output-color=1 might fix that, but it made no difference. I don't think I'm bother about that though.
Thanks again!
